python/pyspark/tests/test_pin_thread.py

```diff
@@ -0,0 +1,156 @@
+            def test_local_property():
+                jvm_thread_id = self.sc._jvm.java.lang.Thread.currentThread().getId()
+                jvm_thread_ids.append(jvm_thread_id)
+
+                # If a property is set in this thread, later it should get the same property
+                # within this thread.
+                self.sc.setLocalProperty(property_name, str(i))
+
+                # Sleep for some arbitrary time.
+                time.sleep(random.random())
+
+                try:
+                    assert self.sc.getLocalProperty(property_name) == str(i)
+
+                    # Each command might create a thread in multi-treading mode in Py4J.
+                    # This assert makes sure that the created thread is being reused.
+                    assert jvm_thread_id == self.sc._jvm.java.lang.Thread.currentThread().getId()
+                except Exception as e:
+                    exceptions.append(e)
+            threads.append(threading.Thread(target=test_local_property))
+
+        for t in threads:
+            t.start()
+
+        for t in threads:
+            t.join()
+
+        for e in exceptions:
+            raise e
+
+        # Created JVM threads should be 10 because Python thread are 10.
+        assert len(set(jvm_thread_ids)) == 10
+
+    def test_multiple_group_jobs(self):
+        # SPARK-22340 Add a mode to pin Python thread into JVM's
+
+        group_a = "job_ids_to_cancel"
+        group_b = "job_ids_to_run"
+
+        threads = []
+        thread_ids = range(4)
+        thread_ids_to_cancel = [i for i in thread_ids if i % 2 == 0]
+        thread_ids_to_run = [i for i in thread_ids if i % 2 != 0]
+
+        # A list which records whether job is cancelled.
+        # The index of the array is the thread index which job run in.
+        is_job_cancelled = [False for _ in thread_ids]
+
+        def run_job(job_group, index):
+            """
+            Executes a job with the group ``job_group``. Each job waits for 3 seconds
+            and then exits.
+            """
+            try:
+                self.sc.setJobGroup(job_group, "test rdd collect with setting job group")
+                self.sc.parallelize([3]).map(lambda x: time.sleep(x)).collect()
+                is_job_cancelled[index] = False
+            except Exception:
+                # Assume that exception means job cancellation.
+                is_job_cancelled[index] = True
+
+        # Test if job succeeded when not cancelled.
+        run_job(group_a, 0)
+        self.assertFalse(is_job_cancelled[0])
+
+        # Run jobs
+        for i in thread_ids_to_cancel:
+            t = threading.Thread(target=run_job, args=(group_a, i))
+            t.start()
+            threads.append(t)
+
+        for i in thread_ids_to_run:
+            t = threading.Thread(target=run_job, args=(group_b, i))
+            t.start()
+            threads.append(t)
+
+        # Wait to make sure all jobs are executed.
+        time.sleep(1)
```

Comment:
I think we can change to barriers & events to avoid sleeps to make this more reliable.  A barrier to check that all jobs have been submitted, and an event to check the jobs have been cancelled.

```python
jobs_submitted = threading.Barrier(num_jobs + 1)
job_cancelled = threading.Event()
...
# in threads that submit jobs
def job_func(x):
  jobs_submitted.wait()
  job_cancelled.wait()
  return x
self.sc.parallelize([3]).map(job_func).collect()
...

# in main thread, after starting all job-submitting threads
jobs_submitted.wait()
self.sc.cancelJobGroup(group_a)
jobs_cancelled.set()
```

Reply:
Seems like `threading.Barrier` is new as of Python 3.2 :-(. I got the point - let me try to fix it

Reply:
Hmmm .. It's kind of tricky to handle it with some kind of minimsed changes ...

Reply:
Ugh thats a headache, but you can work around this, I think -- you can define a holder object which explicitly does not serialize the `Event`, and on un-pickling it'll reset to the shared object. https://stackoverflow.com/a/6313474/1442961

untested, but I think something like:

```python
# module-level global variable -- maybe there is a better way to do this?
job_cancelled = threading.Event()

class EventWrapper:
  event = None
  def __getstate__(self):
    # can't be empty, because then __setstate__ doesn't get called :(
    return {1:2}

  def __setstate__(self, state):
    # ignore the pickled state, just set to the shared event as we're in the same PVM
    self.event = job_cancelled
...

job_cancelled_wrapper = EventWrapper()
def job_func(x):
  ... # something here to replace the barrier ...
  job_cancelled_wrapper.event.wait()
  return x
...
```

I'm surprised this hasn't already come up in pyspark tests.

you're concerned about `Barrier` for spark 2.4,?  because for 3.0, we'll only support 3.4+, right?

Reply:
Oh, no. We will still support python2 in 3.0. It's deprecated in 3.0. Seems like we might depreciate 3.4 and 3.5 too - it's being discussed in the mailing list.

Reply:
Yeah but I thought it's a bit complicated. Given multiple test runs, seems not flaky in Jenkins at least. So I guess it might be fine as is for now. I'll change by using barrier after we release 3.0. Python 2 (and maybe python 3.4 and 3.5) will be dropped at 3.1.

Reply:
oops, sorry I was way off on python support, thanks for correcting me :)

I really think we should make this more reliable, though -- lots of tests have had to be deflaked because of sleeps which usually more than long enough, but fail every so often.  Things which normally take < 1ms have had pauses of 5s (overloaded host machine?  JVM full gc pause?).  A couple of test runs on jenkins here won't find that.

Reply:
Hm, for the suggested codes, I think we won't sill be able to access because `job_func` will be de/serialized in a different PVM.

So, driver side won't be able to access what happens inside the worker sides .. we should open, for instance, another socket or check it, for instance, via writing a file...

Reply:
Seems like Scala side test (`JobCancellationSuite`) is also written with sleep too ..

Reply:
you're right, sorry I was thinking this was like the spark local-mode tests, where everything is in the same JVM.  OK, lets try it this way for now, and then we can change if we it ends up being flaky.

Reply:
Thanks @squito. I increased the timeout to be safer for now.